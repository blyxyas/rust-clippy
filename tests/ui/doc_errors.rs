#![warn(clippy::missing_errors_doc, clippy::doc_markdown)]
#![allow(clippy::result_unit_err)]
#![allow(clippy::unnecessary_wraps)]

use std::io;

pub fn pub_fn_missing_errors_header() -> Result<(), ()> {
    unimplemented!();
}

pub async fn async_pub_fn_missing_errors_header() -> Result<(), ()> {
    unimplemented!();
}

/// This is not sufficiently documented.
pub fn pub_fn_returning_io_result() -> io::Result<()> {
    unimplemented!();
}

/// This is not sufficiently documented.
pub async fn async_pub_fn_returning_io_result() -> io::Result<()> {
    unimplemented!();
}

/// # Errors
/// A description of the errors goes here.
pub fn pub_fn_with_errors_header() -> Result<(), ()> {
    unimplemented!();
}

/// # Errors
/// A description of the errors goes here.
pub async fn async_pub_fn_with_errors_header() -> Result<(), ()> {
    unimplemented!();
}

/// This function doesn't require the documentation because it is private
fn priv_fn_missing_errors_header() -> Result<(), ()> {
    unimplemented!();
}

/// This function doesn't require the documentation because it is private
async fn async_priv_fn_missing_errors_header() -> Result<(), ()> {
    unimplemented!();
}

pub struct Struct1;

impl Struct1 {
    /// This is not sufficiently documented.
    pub fn pub_method_missing_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /// This is not sufficiently documented.
    pub async fn async_pub_method_missing_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /// # Errors
    /// A description of the errors goes here.
    pub fn pub_method_with_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /**
    Before text
    ```
        Indentation block
    ```
    # Errors
    Hi
     */
    pub fn pub_method_with_block_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /**
    This is not sufficiently documented.
    */
    pub fn pub_method_missing_block_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /**
    # Errors
    This is sufficiently documented.

    This function is a test for this lint's capabilities with other
    lints, such as `doc_markdown`
    std::str;
    */
    pub fn pub_method_block_errors_header_doc_markdown() -> Result<(), ()> {
        unimplemented!();
    }

    /**
    This is not sufficiently documented.

    This function is also a test for this lint's capabilities with other
    lints, such as `doc_markdown`
    std::str;
    */
    pub fn pub_method_missing_block_errors_header_doc_markdown() -> Result<(), ()> {
        unimplemented!();
    }

    /// # Errors
    /// A description of the errors goes here.
    pub async fn async_pub_method_with_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /// This function doesn't require the documentation because it is private.
    fn priv_method_missing_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    /// This function doesn't require the documentation because it is private.
    async fn async_priv_method_missing_errors_header() -> Result<(), ()> {
        unimplemented!();
    }
}

pub trait Trait1 {
    /// This is not sufficiently documented.
    fn trait_method_missing_errors_header() -> Result<(), ()>;

    /// # Errors
    /// A description of the errors goes here.
    fn trait_method_with_errors_header() -> Result<(), ()>;
}

impl Trait1 for Struct1 {
    fn trait_method_missing_errors_header() -> Result<(), ()> {
        unimplemented!();
    }

    fn trait_method_with_errors_header() -> Result<(), ()> {
        unimplemented!();
    }
}

fn main() -> Result<(), ()> {
    Ok(())
}
