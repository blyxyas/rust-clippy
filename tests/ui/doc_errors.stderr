error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:7:1
   |
LL | pub fn pub_fn_missing_errors_header() -> Result<(), ()> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::missing-errors-doc` implied by `-D warnings`

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:11:1
   |
LL | pub async fn async_pub_fn_missing_errors_header() -> Result<(), ()> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:16:1
   |
LL | pub fn pub_fn_returning_io_result() -> io::Result<()> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:21:1
   |
LL | pub async fn async_pub_fn_returning_io_result() -> io::Result<()> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:51:5
   |
LL |     pub fn pub_method_missing_errors_header() -> Result<(), ()> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:56:5
   |
LL |     pub async fn async_pub_method_missing_errors_header() -> Result<(), ()> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:81:5
   |
LL |     pub fn pub_method_missing_block_errors_header() -> Result<(), ()> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:104:5
   |
LL |     pub fn pub_method_missing_block_errors_header_doc_markdown() -> Result<(), ()> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: docs for function returning `Result` missing `# Errors` section
  --> $DIR/doc_errors.rs:127:5
   |
LL |     fn trait_method_missing_errors_header() -> Result<(), ()>;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 9 previous errors

